robyn/reloader.py

```diff
@@ -0,0 +1,84 @@
+import os
+import signal
+import subprocess
+import sys
+import time
+import psutil
+
+from watchdog.events import FileSystemEventHandler
+from watchdog.observers import Observer
+
+from robyn.logger import Colors, logger
+
+
+def setup_reloader(directory_path: str, file_path: str):
+    event_handler = EventHandler(file_path)
+
+    event_handler.reload()
+
+    logger.info(
+        f"Dev server initialized with the directory_path : {directory_path}",
+        Colors.BLUE,
+    )
+
+    def terminating_signal_handler(_sig, _frame):
+        event_handler.stop_server()
+        logger.info("Terminating reloader", bold=True)
+        observer.stop()
+        observer.join()
+
+    signal.signal(signal.SIGINT, terminating_signal_handler)
+    signal.signal(signal.SIGTERM, terminating_signal_handler)
+
+    observer = Observer()
+    observer.schedule(event_handler, path=directory_path, recursive=True)
+    observer.start()
+
+    try:
+        while observer.is_alive():
+            observer.join(1)
+    finally:
+        observer.stop()
+        observer.join()
+
+
+class EventHandler(FileSystemEventHandler):
+    def __init__(self, file_path: str) -> None:
+        self.file_path = file_path
+
+        self.last_reload = time.time()
+
+    def stop_server(self):
+        for process in psutil.Process().children(recursive=True):
+            process.kill()
+            process.wait()  # Required to avoid zombies
+
+    def reload(self):
+        self.stop_server()
+
+        new_env = os.environ.copy()
+        new_env[
+            "IS_RELOADER_SETUP"
+        ] = "True"  # This is used to check if a reloader is already running
```

Comment:
Hey @AntoineRR ,

All looks great. I just have one final question - does `IS_RELOADER_SETUP` env variable indicate that it is a subprocess generated from a change ?

Reply:
This env variable is used to check if the reloader is running. You can think of it as a check that the subprocess was launched after a file modification by the reloader yes

Reply:
@AntoineRR , but why do we need an environment variable for this? Just for sharing data between processes?

We can use `multiprocessing.Event` or `Lock` for the same - https://docs.python.org/3/library/multiprocessing.html?highlight=event#multiprocessing.Event

Reply:
I am not sure we can use this to share information between processes, but only between threads from the same process? I based my implementation on an old commit from sanic, but maybe they improved the way it is done since -> https://github.com/sanic-org/sanic/commit/52c2a8484e6aa5fa13aaade49e1f2597dd006e15.
If you have an alternative in mind please let me know, as I don't see how to use a `Lock` or `Event` in this situation.